### PR TITLE
Add initial frontend development documentation

### LINKED ISSUE #129: Document frontend development workflow
**Summary:** 

Frontend development workflow should be more thoroughly documented
References https://github.com/TheThingsNetwork/lorawan-stack/pull/109#discussion_r255879919

**Why do we need this?**

To give our users proper development documentation

**What is already there? What do you see now?**

Nothing

**What is missing? What do you want to see?**

Documented frontend workflow

**How do you propose to implement this?**

@bafonins or @kschiffer should write down what they do to develop console

### PR DESCRIPTION
**Pull request by kschiffer**
#### Summary
Closes #129.
This PR provides a proper starting point for the development documentation of the frontend.

#### Changes
- Added frontend documentation to `DEVELOPMENT.md`

#### Notes for Reviewers
I hope this is sufficient for a start. I likely forgot some details so please look through it and notify me if you miss anything.
I think I should follow up with a couple of improvements to the build tooling:
- Assume `NODE_ENV=production` by default (is currently `development`)
- Add either another target `make dev.init.stack-dev` or add `--redirect-uri 'http://localhost:8080/console/oauth/callback'` to `make dev.init.stack` to make sure the OAuth flow works properly through the frontend development server

### CODE REVIEW

**Comment by htdvisser** [comment]
> Assume NODE_ENV=production by default (is currently development)

Yes. Please add that.

> Add either another target make dev.init.stack-dev or add --redirect-uri 'http://localhost:8080/console/oauth/callback' to make dev.init.stack

I think it would be better to add a list of development-specific environment to `DEVELOPMENT.md`. Then we can set `TTN_LW_HTTP_LISTEN=":1885"` to a different port and serve the frontend dev server (including proxy) on 1885.

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -166,15 +166,118 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
 ├── public              frontend code will be compiled to this folder - not added to git
 ├── release             binaries will be compiled to this folder - not added to git
 └── sdk                 source code for our SDKs
     └── js              source code for our JavaScript SDK
 ```
+
+
+## Frontend
+### Introduction
+The TTN stack for LoRaWAN employs two frontend applications: the **Console** and **OAuth Provider**. Both applications use [React](https://reactjs.org/) as frontend framework. The `console` package of the backend exposes the webserver and – together with the `oauth` package – is in charge of handling authorization logic (OAuth flow, login/logout).
```
Comment (htdvisser) [inline DEVELOPMENT.md]: `The TTN stack` -> `The Things Network Stack`, right, @johanstokking? We do this wrong in more places in this document.

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -166,15 +166,118 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+
+#### Console
+The Console is the official management application of the stack. It can be used to register applications, devices or gateways, to monitor network traffic, or configuring network related options, among other things. It is a single page application (SPA) that runs entirely in the browser and uses an access token to communicate with the stack.
```
Comment (htdvisser) [inline DEVELOPMENT.md]: `devices` -> `end devices`
`access token` -> `OAuth access token`

File: DEVELOPMENT.md [hunk 3]
```diff
@@ -166,15 +166,118 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+
+#### OAuth
+The OAuth app provides the necessary frontend for the OAuth provider of the stack. It is used e.g. to display the authorization screen that users get prompted with when they want to authorize a third-party app to access the stack.
+
+### Building the frontend
+You can control whether to build the frontend for production or development by setting the `$NODE_ENV` environment variable to either `development` or `production`. The frontend is then build using:
```
Comment (htdvisser) [inline DEVELOPMENT.md]: `then build using` -> `can then be built with`

File: DEVELOPMENT.md [hunk 4]
```diff
@@ -166,15 +166,118 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+
+```sh
+mage js:build
+```
+*Note that the stack needs to be initialized first using `mage init`*
```
Comment (htdvisser) [inline DEVELOPMENT.md]: This is already explained on L34

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -166,15 +166,118 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+After adding messages this way, it needs to be added the locales file `pkg/webui/locales/*.js` by using:
+
+```sh
+mage js:translations
+```
+*Note: When using `mage js:serve`, this command will be run automatically after any change*
+
+The message definitions in `pkg/webui/locales` can be used to provide translations in other languages (e.g. `fr.js`). Keep in mind that locale files are checked in and commited, any discrepancy in the locales file with the defined messages will lead to a CI failure.
+
+### Folder Structure
```
Comment (htdvisser) [inline DEVELOPMENT.md]: Frontend Folder Structure (we already have a `#folder-structure` above)

File: DEVELOPMENT.md (hunk 1 above)
Comment (kschiffer) [reply DEVELOPMENT.md]: Ah yeah, I was actually writing it this way, but then changed it after I saw it was written differently in the rest of the file.

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -166,15 +166,118 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+* Build the JS SDK
+* Extract backend locale messages
+* Install frontend node dependencies via `yarn`
+* Build the frontend (using `webpack`) and output into `/public`
+
+The difference of a development build includes:
+
+* Including source maps
+* Using DLL bundle for modules to reduce build time
+* Means to stub backend data to make `webpack-dev-server` work properly
```
Comment (bafonins) [inline DEVELOPMENT.md]: `webpack-dev-server` is used proxy request from the console to the stack, no stubs.

File: DEVELOPMENT.md [hunk 7]
```diff
@@ -166,15 +166,118 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+
+After successfully running the build command, the stack has all necessary files to run the Console and OAuth provider applications.
+
+### Development
+#### Serving the frontend for development
+For development purposes, the frontend can be run using `webpack-dev-server`. After following the [Getting Started](#getting-started) section to initialize the stack and doing an initial build of the frontend via `mage js:build`, it can be served using:
+
+```sh
+mage js:serve
```
Comment (bafonins) [inline DEVELOPMENT.md]: i would include `export NODE_ENV=development`

File: DEVELOPMENT.md [hunk 8]
```diff
@@ -166,15 +166,118 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+```
+*Make sure `$NODE_ENV` equals `development`*
```
Comment (bafonins) [inline DEVELOPMENT.md]: see above

**Comment by bafonins** [comment]
I think it also worth mentioning the flags we use for `ttn-lw-stack` when working in the dev mode

**Comment by htdvisser** [comment]
I would strongly suggest to use environment instead of flags. That way people can just put everything in their `.envrc` and have `direnv` load it when they enter the directory

File: DEVELOPMENT.md (hunk 1 above)
Comment (kschiffer) [reply DEVELOPMENT.md]: My suggestion would also be to put the brand and stack name always in italics in our docs, makes it easier to scan and less clashing with other parts of the sentence: notice `The The` in `The The Things Network Stack employs …`.

**Comment by kschiffer** [comment]
> I think it would be better to add a list of development-specific environment to `DEVELOPMENT.md`. Then we can set `TTN_LW_HTTP_LISTEN=":1885"` to a different port and serve the frontend dev server (including proxy) on 1885.

I actually like the fact that we do development on port 8080. It's a good visual clue to know what config is currently running.
I've added the environment setup for development to the doc.

File: DEVELOPMENT.md (hunk 1 above)
Comment (htdvisser) [reply DEVELOPMENT.md]: There should not be a double "The", it's "The Things Network Stack" or "the stack".

File: DEVELOPMENT.md [hunk 9]
```diff
@@ -166,15 +166,131 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+export NODE_ENV=development
+mage js:serve
+```
+
+The development server runs on `http://localhost:8080` and will proxy all api calls to port `1885`. The serve command watches any changes inside `pkg/webui` and refreshes automatically. 
+In order to set up the stack to support running the frontend via `webpack-dev-server`, the following environment setup is needed:
+
+```
+NODE_ENV=development
+LOG_LEVEL=debug
```
Comment (htdvisser) [inline DEVELOPMENT.md]: These all miss the `TTN_LW` prefix.

File: DEVELOPMENT.md [hunk 10]
```diff
@@ -166,15 +166,131 @@ The actual commands for compilation are handled by our Makefile, so the only thi
...
+IS_OAUTH_UI_JS_FILE=libs.bundle.js,oauth.js
+CONSOLE_UI_JS_FILE=libs.bundle.js,console.js
+CONSOLE_UI_CANONICAL_URL=http://localhost:8080/console
+CONSOLE_OAUTH_AUTHORIZE_URL=http://localhost:8080/oauth/authorize
+CONSOLE_OAUTH_TOKEN_URL=http://localhost:8080/oauth/token
+IS_OAUTH_UI_CANONICAL_URL=http://localhost:8080/oauth
+IS_EMAIL_NETWORK_IDENTITY_SERVER_URL=http://localhost:8080/oauth.js
```
Comment (bafonins) [inline DEVELOPMENT.md]: `TTN_LW_CONSOLE_UI_ASSETS_BASE_URL=http://localhost:8080/assets`

**Comment by bafonins** [comment]
good! i think it is a good start

**Pull request merged**